### PR TITLE
MGMT-18292: import openshift installer and using it to generate   configuration iso

### PR DESCRIPTION
**Pull request by tsorya**

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -18,11 +18,14 @@ COPY vendor/ vendor/
...
 RUN GODEBUG=madvdontneed=1 GOGC=50 CGO_ENABLED=1 GOOS=${TARGETOS:-linux} GOARCH=${TARGETARCH} go build -a -o build/server cmd/server/main.go
 
-FROM registry.access.redhat.com/ubi9/ubi-minimal:9.4
+
+FROM registry.ci.openshift.org/ocp/4.17:base-rhel9
 
 ARG DATA_DIR=/data
 RUN mkdir $DATA_DIR && chmod 775 $DATA_DIR
 
+RUN dnf install -y nmstate-libs nmstate && dnf clean all && rm -rf /var/cache/dnf/*
```
Comment (eranco74) [inline Dockerfile:27]: Specify the nmstate version you wish to install, this effects the nmstate API we support (e.g. mac-identifier) so we should probably document it)

File: api/v1alpha1/imageclusterinstall_types.go [hunk 2]
```diff
@@ -100,6 +100,11 @@ type ImageClusterInstallSpec struct {
 	// Proxy defines the proxy settings to be applied in relocated cluster
 	// +optional
 	Proxy *Proxy `json:"proxy,omitempty"`
+
+	// AdditionalNTPSources is a list of NTP sources (hostname or IP) to be added to all cluster
+	// hosts. They are added to any NTP sources that were configured through other means.
+	// +optional
+	AdditionalNTPSources []string `json:"additionalNTPSources,omitempty"`
```
Comment (eranco74) [inline api/v1alpha1/imageclusterinstall_types.go:107]: :+1: 
I didn't know we (does recert?) support this

File: go.mod [hunk 3]
```diff
@@ -1,101 +1,250 @@
...
+	k8s.io/api v0.31.1
+	k8s.io/apimachinery v0.31.1
+	k8s.io/client-go v0.31.1
+	k8s.io/utils v0.0.0-20240921022957-49e7df575cb6
+	sigs.k8s.io/controller-runtime v0.19.0
+	sigs.k8s.io/yaml v1.4.0
 )
 
 require (
+	cloud.google.com/go/auth v0.7.2 // indirect
```
Comment (eranco74) [inline go.mod:33]: Wow what a nightmare

File: go.mod [hunk 4]
```diff
@@ -1,101 +1,250 @@
...
-	sigs.k8s.io/yaml v1.4.0 // indirect
 )
 
 replace github.com/metal3-io/baremetal-operator/apis => github.com/openshift/baremetal-operator/apis v0.0.0-20240611081634-a92cc4b9ae8b
 
-replace k8s.io/client-go => k8s.io/client-go v0.29.4
+replace (
+	github.com/openshift/assisted-service/api => github.com/openshift/assisted-service/api v0.0.0-20241001055825-63e8b0d3ad63
+	github.com/openshift/assisted-service/client => github.com/openshift/assisted-service/client v0.0.0-20241001055825-63e8b0d3ad63
+	github.com/openshift/assisted-service/models => github.com/openshift/assisted-service/models v0.0.0-20241001055825-63e8b0d3ad63
```
Comment (eranco74) [inline go.mod:246]: Do we need all 3?

File: internal/installer/installer.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+	"github.com/openshift/installer/pkg/asset"
+	"github.com/openshift/installer/pkg/asset/imagebased/configimage"
+	"github.com/openshift/installer/pkg/asset/kubeconfig"
+	"github.com/openshift/installer/pkg/asset/password"
+	assetStore "github.com/openshift/installer/pkg/asset/store"
+	"github.com/sirupsen/logrus"
+)
+
+//go:generate mockgen --build_flags=--mod=mod -package=installer -destination=mock_installer.go . IInstaller
+type IInstaller interface {
```
Comment (eranco74) [inline internal/installer/installer.go:15]: Should be Installer

File: internal/installer/installer.go [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+		&configimage.ConfigImage{},
+		&kubeconfig.ImageBasedAdminClient{},
+		&password.KubeadminPassword{},
+	}}
+}
+
+func (i *installer) CreateInstallationIso(ctx context.Context, log logrus.FieldLogger, workDir string) error {
+	log.Info("Creating installation ISO from %s", workDir)
+	fetcher := assetStore.NewAssetsFetcher(workDir)
+	return fetcher.FetchAndPersist(ctx, i.assets)
```
Comment (eranco74) [inline internal/installer/installer.go:34]: Nice

File: api/v1alpha1/imageclusterinstall_types.go (hunk 2 above)
Comment (tsorya) [reply api/v1alpha1/imageclusterinstall_types.go:107]: Yeah, we support if https://github.com/openshift-kni/lifecycle-agent/blob/main/api/seedreconfig/seedreconfig.go#L147

File: go.mod (hunk 3 above)
Comment (tsorya) [reply go.mod:33]: totally

File: go.mod (hunk 4 above)
Comment (tsorya) [reply go.mod:246]: yep as this is what installer inspects

Comment (tsorya) [reply go.mod:246]: installer expects for them so sort of yes

File: cmd/manager/main.go [hunk 7]
```diff
@@ -158,10 +157,10 @@ func main() {
 		Scheme:          mgr.GetScheme(),
 		Options:         controllerOptions,
 		BaseURL:         baseURL,
-		CertManager:     certs.KubeConfigCertManager{},
 		NoncachedClient: mgr.GetAPIReader(),
+		Installer:       installer.NewInstaller(),
 	}).SetupWithManager(mgr); err != nil {
-		setupLog.Error(err, "unable to create controller", "controller", "ImageClusterInstall")
+		setupLog.Error(err, "unable to create controller", "controller", "ImageClusterInstallMonitor")
```
Comment (eranco74) [inline cmd/manager/main.go:163]: Should say: `ImageClusterInstall`

File: controllers/imageclusterinstall_controller.go [hunk 8]
```diff
@@ -690,47 +697,37 @@ func (r *ImageClusterInstallReconciler) labelReferencedObjectsForBackup(ctx cont
...
-func (r *ImageClusterInstallReconciler) clusterInfoFilePath(ici *v1alpha1.ImageClusterInstall) (string, error) {
-	_, filesDir, err := r.configDirs(ici)
-	if err != nil {
-		return "", err
-	}
-
-	return filepath.Join(filesDir, clusterConfigDir, "manifest.json"), nil
-}
-
 // writeInputData writes the required info based on the ImageClusterInstall to the config cache dir
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:713]: Update the doc string

File: controllers/imageclusterinstall_controller.go [hunk 9]
```diff
@@ -690,47 +697,37 @@ func (r *ImageClusterInstallReconciler) labelReferencedObjectsForBackup(ctx cont
...
 	}
 
-	locked, lockErr, funcErr := filelock.WithWriteLock(lockDir, func() error {
+	locked, lockErr, funcErr := filelock.WithWriteLock(lockDir, func() (err error) {
 
-		manifestsPath := filepath.Join(clusterConfigPath, manifestsDir)
-		if err := os.MkdirAll(manifestsPath, 0700); err != nil {
+		clusterConfigPath := filepath.Join(filesDir, ClusterConfigDir)
+		if verifyIsoAndAuthExists(clusterConfigPath) {
+			log.Infof("config iso already exists, skipping")
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:729]: Remove this log and add a log saying that we should generate them in line 728.

File: controllers/imageclusterinstall_controller.go [hunk 10]
```diff
@@ -739,22 +736,24 @@ func (r *ImageClusterInstallReconciler) writeInputData(
...
+			return fmt.Errorf("failed to get ca bundle: %w", err)
+		}
+
+		extraManifestsPath := filepath.Join(clusterConfigPath, extraManifestsDir)
+		if err := os.MkdirAll(extraManifestsPath, 0700); err != nil {
+			return err
 		}
-		if err := r.writeInvokerCM(filepath.Join(manifestsPath, invokerCMFileName)); err != nil {
+
+		if err := r.writeInvokerCM(filepath.Join(extraManifestsPath, invokerCMFileName)); err != nil {
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:755]: Extract this to another function that handles all the extraManifests

File: controllers/imageclusterinstall_controller.go [hunk 11]
```diff
@@ -902,48 +885,72 @@ func (r *ImageClusterInstallReconciler) writeClusterInfo(ctx context.Context, lo
...
+		return fmt.Errorf("failed to marshal cluster info: %w", err)
+	}
+	if err := os.WriteFile(file, data, 0644); err != nil {
+		return fmt.Errorf("failed to write cluster info: %w", err)
 	}
 
+	return nil
+}
+
+func (r *ImageClusterInstallReconciler) writeInstallConfig(
```
Comment (eranco74) [inline controllers/imageclusterinstall_controller.go:920]: Make this a static method, it doesn't require the ImageClusterInstallReconciler.
Also consider moving it to the installer package (seems to belong there anyway)

**Review by eranco74 — COMMENTED** [review]
looks good, just nits

**Pull request merged**